### PR TITLE
Fix memory leak in Find in Files

### LINKED ISSUE #31570: Memory leak when using Find in Files
**Godot version:**
3.2.dev.custom_build.05a431089

**Issue description:**
```
Direct leak of 6768 byte(s) in 94 object(s) allocated from:
    #0 0x7f78d9fb7448 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10c448)
    #1 0xe68dee5 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:85
    #2 0xe68ddc8 in operator new(unsigned long, char const*) core/os/memory.cpp:42
    #3 0x4d6276d in FileAccess* FileAccess::_create_builtin<FileAccessUnix>() core/os/file_access.h:77
    #4 0xe5b65ea in FileAccess::create(FileAccess::AccessType) core/os/file_access.cpp:49
    #5 0xe5b6a29 in FileAccess::create_for_path(String const&) core/os/file_access.cpp:76
    #6 0xe5b6d90 in FileAccess::open(String const&, int, Error*) core/os/file_access.cpp:108
    #7 0x89d0997 in FindInFiles::_scan_file(String) editor/find_in_files.cpp:256
    #8 0x89cecde in FindInFiles::_iterate() editor/find_in_files.cpp:204
    #9 0x89cd8ff in FindInFiles::_process() editor/find_in_files.cpp:158
    #10 0x89cbbcd in FindInFiles::_notification(int) editor/find_in_files.cpp:117
    #11 0x89fef1f in FindInFiles::_notificationv(int, bool) editor/find_in_files.h:39
    #12 0xe117a6d in Object::notification(int, bool) core/object.cpp:931
    #13 0x90b842b in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:957
    #14 0x90a90d6 in SceneTree::idle(float) scene/main/scene_tree.cpp:515
    #15 0x1530ae8 in Main::iteration() main/main.cpp:1920
    #16 0x142ff98 in OS_X11::run() platform/x11/os_x11.cpp:3181
    #17 0x13acb93 in main platform/x11/godot_x11.cpp:56
    #18 0x7f78d86bbb6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)

Direct leak of 4368 byte(s) in 78 object(s) allocated from:
    #0 0x7f78d9fb7448 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10c448)
    #1 0xe68dee5 in Memory::alloc_static(unsigned long, bool) core/os/memory.cpp:85
    #2 0xe68ddc8 in operator new(unsigned long, char const*) core/os/memory.cpp:42
    #3 0x4d627cb in DirAccess* DirAccess::_create_builtin<DirAccessUnix>() core/os/dir_access.h:65
    #4 0xe5af6b6 in DirAccess::create(DirAccess::AccessType) core/os/dir_access.cpp:261
    #5 0xe5aefdf in DirAccess::create_for_path(String const&) core/os/dir_access.cpp:231
    #6 0xe5af0ef in DirAccess::open(String const&, Error*) core/os/dir_access.cpp:245
    #7 0x89cfe23 in FindInFiles::_scan_dir(String, PoolVector<String>&) editor/find_in_files.cpp:224
    #8 0x89ce34c in FindInFiles::_iterate() editor/find_in_files.cpp:182
    #9 0x89cd8ff in FindInFiles::_process() editor/find_in_files.cpp:158
    #10 0x89cbbcd in FindInFiles::_notification(int) editor/find_in_files.cpp:117
    #11 0x89fef1f in FindInFiles::_notificationv(int, bool) editor/find_in_files.h:39
    #12 0xe117a6d in Object::notification(int, bool) core/object.cpp:931
    #13 0x90b842b in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:957
    #14 0x90a90d6 in SceneTree::idle(float) scene/main/scene_tree.cpp:515
    #15 0x1530ae8 in Main::iteration() main/main.cpp:1920
    #16 0x142ff98 in OS_X11::run() platform/x11/os_x11.cpp:3181
    #17 0x13acb93 in main platform/x11/godot_x11.cpp:56
    #18 0x7f78d86bbb6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)

Direct leak of 73 byte(s) in 1 object(s) allocated from:
    #0 0x7f78d9fb7448 in malloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10c448)
    #1 0x7f78d9970064 in XGetWindowProperty (/lib/x86_64-linux-gnu/libX11.so.6+0x24064)

Indirect leak of 39392 byte(s) in 94 object(s) allocated from:
    #0 0x7f78d9fb787e in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10c87e)
    #1 0xe68eb3e in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x316ea46 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x53988db in String::resize(int) core/ustring.h:154
    #4 0xe342517 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xe35d704 in String::replace(String const&, String const&) const core/ustring.cpp:2975
    #6 0xe5b7527 in FileAccess::fix_path(String const&) const core/os/file_access.cpp:142
    #7 0x53d3a47 in FileAccessUnix::_open(String const&, int) drivers/unix/file_access_unix.cpp:76
    #8 0xe5b6f4b in FileAccess::open(String const&, int, Error*) core/os/file_access.cpp:109
    #9 0x89d0997 in FindInFiles::_scan_file(String) editor/find_in_files.cpp:256
    #10 0x89cecde in FindInFiles::_iterate() editor/find_in_files.cpp:204
    #11 0x89cd8ff in FindInFiles::_process() editor/find_in_files.cpp:158
    #12 0x89cbbcd in FindInFiles::_notification(int) editor/find_in_files.cpp:117
    #13 0x89fef1f in FindInFiles::_notificationv(int, bool) editor/find_in_files.h:39
    #14 0xe117a6d in Object::notification(int, bool) core/object.cpp:931
    #15 0x90b842b in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:957
    #16 0x90a90d6 in SceneTree::idle(float) scene/main/scene_tree.cpp:515
    #17 0x1530ae8 in Main::iteration() main/main.cpp:1920
    #18 0x142ff98 in OS_X11::run() platform/x11/os_x11.cpp:3181
    #19 0x13acb93 in main platform/x11/godot_x11.cpp:56
    #20 0x7f78d86bbb6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)

Indirect leak of 26080 byte(s) in 78 object(s) allocated from:
    #0 0x7f78d9fb787e in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10c87e)
    #1 0xe68eb3e in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x316ea46 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x53988db in String::resize(int) core/ustring.h:154
    #4 0xe342517 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xe3421c6 in String::operator+(String const&) const core/ustring.cpp:252
    #6 0xe35e177 in String::replace_first(String const&, String const&) const core/ustring.cpp:3010
    #7 0xe5ae869 in DirAccess::fix_path(String) const core/os/dir_access.cpp:194
    #8 0x53ceb79 in DirAccessUnix::change_dir(String) drivers/unix/dir_access_unix.cpp:299
    #9 0xe5af350 in DirAccess::open(String const&, Error*) core/os/dir_access.cpp:248
    #10 0x89cfe23 in FindInFiles::_scan_dir(String, PoolVector<String>&) editor/find_in_files.cpp:224
    #11 0x89ce34c in FindInFiles::_iterate() editor/find_in_files.cpp:182
    #12 0x89cd8ff in FindInFiles::_process() editor/find_in_files.cpp:158
    #13 0x89cbbcd in FindInFiles::_notification(int) editor/find_in_files.cpp:117
    #14 0x89fef1f in FindInFiles::_notificationv(int, bool) editor/find_in_files.h:39
    #15 0xe117a6d in Object::notification(int, bool) core/object.cpp:931
    #16 0x90b842b in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:957
    #17 0x90a90d6 in SceneTree::idle(float) scene/main/scene_tree.cpp:515
    #18 0x1530ae8 in Main::iteration() main/main.cpp:1920
    #19 0x142ff98 in OS_X11::run() platform/x11/os_x11.cpp:3181
    #20 0x13acb93 in main platform/x11/godot_x11.cpp:56
    #21 0x7f78d86bbb6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)

Indirect leak of 23264 byte(s) in 94 object(s) allocated from:
    #0 0x7f78d9fb787e in __interceptor_realloc (/lib/x86_64-linux-gnu/libasan.so.5+0x10c87e)
    #1 0xe68eb3e in Memory::realloc_static(void*, unsigned long, bool) core/os/memory.cpp:140
    #2 0x316ea46 in CowData<wchar_t>::resize(int) core/cowdata.h:282
    #3 0x53988db in String::resize(int) core/ustring.h:154
    #4 0xe342517 in String::operator+=(String const&) core/ustring.cpp:276
    #5 0xe3421c6 in String::operator+(String const&) const core/ustring.cpp:252
    #6 0xe3708d6 in String::plus_file(String const&) const core/ustring.cpp:3948
    #7 0x89d05ab in FindInFiles::_scan_dir(String, PoolVector<String>&) editor/find_in_files.cpp:248
    #8 0x89ce34c in FindInFiles::_iterate() editor/find_in_files.cpp:182
    #9 0x89cd8ff in FindInFiles::_process() editor/find_in_files.cpp:158
    #10 0x89cbbcd in FindInFiles::_notification(int) editor/find_in_files.cpp:117
    #11 0x89fef1f in FindInFiles::_notificationv(int, bool) editor/find_in_files.h:39
    #12 0xe117a6d in Object::notification(int, bool) core/object.cpp:931
    #13 0x90b842b in SceneTree::_notify_group_pause(StringName const&, int) scene/main/scene_tree.cpp:957
    #14 0x90a90d6 in SceneTree::idle(float) scene/main/scene_tree.cpp:515
    #15 0x1530ae8 in Main::iteration() main/main.cpp:1920
    #16 0x142ff98 in OS_X11::run() platform/x11/os_x11.cpp:3181
    #17 0x13acb93 in main platform/x11/godot_x11.cpp:56
    #18 0x7f78d86bbb6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)
```

### PR DESCRIPTION
**Pull request by ghost**
I have implemented a fix for memory leak by hint of @Zylann in issue #31570

_Bugsquad edit: Fix #31570_

### CODE REVIEW

File: editor/find_in_files.cpp [hunk 1]
```diff
@@ -221,7 +221,7 @@ float FindInFiles::get_progress() const {
 
 void FindInFiles::_scan_dir(String path, PoolStringArray &out_folders) {
 
-	DirAccess *dir = DirAccess::open(path);
+	DirAccessRef *dir = DirAccessRef::open(path);
```
Comment (Zylann) [inline editor/find_in_files.cpp]: `DirAccessRef` is a wrapper to a pointer, there is no need to use a pointer type here

File: editor/find_in_files.cpp [hunk 2]
```diff
@@ -221,16 +221,16 @@ float FindInFiles::get_progress() const {
...
-	DirAccessRef *dir = DirAccessRef::open(path);
-	if (dir == NULL) {
+	DirAccessRef dir = DirAccess::open(path);
+	if (dir) {
 		print_verbose("Cannot open directory! " + path);
 		return;
 	}
 
-	dir->list_dir_begin();
+	dir.list_dir_begin();
```
Comment (Zylann) [inline editor/find_in_files.cpp]: This won't compile either. It has to remain `->`. DirAccessRef is a wrapper to pointer, you don't need to change this. Please try to compile Godot to be sure your change works.

File: editor/find_in_files.cpp [hunk 3]
```diff
@@ -221,8 +221,8 @@ float FindInFiles::get_progress() const {
 
 void FindInFiles::_scan_dir(String path, PoolStringArray &out_folders) {
 
-	DirAccess *dir = DirAccess::open(path);
-	if (dir == NULL) {
+	DirAccessRef dir = DirAccess::open(path);
+	if (dir) {
```
Comment (akien-mga) [inline editor/find_in_files.cpp]: That's wrong, `if (dir)` is like `if (dir != NULL)`, which is the opposite of the previous code. That will break the editor.

Comment (akien-mga) [reply editor/find_in_files.cpp]: Literally the only change needed here was to replace `DirAccess *` by `DirAccessRef`, which is equivalent but freed when out of scope, or add a `memdelete(dir)`  before any return.

**Comment by akien-mga** [comment]
Eventually you should squash the commits together into one, and edit the resulting commit log to be more explicit about what the commit does (fixing a memory leak). Mentioning the GitHub issue being fixed is good, but it's not sufficient when reading the changelog (`git log`).

See https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-logs-with-readability-in-mind

**Comment by akien-mga** [comment]
Thanks!

**Comment by Klesomik** [comment]
Thank you so much! 
I've learned a lot of new things by your replies. 
Your community is really firendly to newbies!

**Pull request merged**